presto-tests/src/test/java/com/facebook/presto/tests/TestQueryManager.java

```diff
@@ -90,6 +92,39 @@ public void testFailQuery()
         assertEquals(queryInfo.getErrorCode(), GENERIC_INTERNAL_ERROR.toErrorCode());
         assertNotNull(queryInfo.getFailureInfo());
         assertEquals(queryInfo.getFailureInfo().getMessage(), "mock exception");
+        assertEquals(0, queryManager.getStats().getQueuedQueries());
+    }
+
+    @Test(timeOut = 60_000L)
+    public void testFailQueryPrerun()
+            throws Exception
+    {
+        DispatchManager dispatchManager = queryRunner.getCoordinator().getDispatchManager();
+        QueryId queryId = dispatchManager.createQueryId();
+        dispatchManager.createQuery(
+                queryId,
+                "slug",
+                new TestingSessionContext(TEST_SESSION),
+                "SELECT * FROM lineitem")
+                .get();
+
+        // wait until it's admitted but fail it before it starts
+        while (true) {
+            QueryState state = dispatchManager.getQueryInfo(queryId).getState();
+            if (state.ordinal() >= RUNNING.ordinal()) {
+                fail("unexpected query state: " + state);
+            }
+            if (state.ordinal() >= DISPATCHING.ordinal()) {
+                System.out.println("Breaking on state:" + state);
+                break;
+            }
+            Thread.sleep(10);
+        }
+
+        // cancel query
+        QueryManager queryManager = queryRunner.getCoordinator().getQueryManager();
+        queryManager.failQuery(queryId, new PrestoException(GENERIC_USER_ERROR, "mock exception"));
```

Comment:
Also add a check here to make sure we're not killing an already-running query.

Reply:
Not sure why that should be necessary; the test will fail on line 115 if it gets as far as RUNNING

Reply:
Query execution is proceeding in the background, so this is a last-moment check that the presumptions still hold.

Reply:
Fair enough